### PR TITLE
Add is_unique function and corresponding test suite

### PR DESCRIPTION
**Pull request by ciiyaa**
---
name: solution review 
about: A template PR for code review with a checklist
---

The `is_unique` function checks if all characters in a given string are unique.

## Behavior

Accepts a string and returns True if all characters are unique and False otherwise. Spaces and special characters are considered valid, and the check is case-sensitive.

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations
- [ ] The function is not called in the function file

## Strategy

The function first checks if the input is a string, then compares the length of the string with the length of a set created from it. If the lengths are equal, it returns `True ` (all characters are unique), otherwise `False`

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_is_unique.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        """It should return True for an empty string."""
+        self.assertTrue(is_unique(""))
+
+    def test_space_character(self):
+        """It should return True for a string with a single space."""
+        self.assertTrue(is_unique(" "))
+
+    # Edge cases
+    def test_mixed_cases(self):
+        """It should return True for a string with mixed cases."""
```
Comment (Urz1) [inline solutions/tests/test_is_unique.py:46]: It might be worth clarifying the test case description to emphasize the case-sensitive behavior.
```suggestion
        """It should return True for a case-sensitive string with unique mixed-case characters."""
```

File: solutions/tests/test_is_unique.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        self.assertTrue(is_unique("aAbBcC"))
+
+    def test_special_characters(self):
+        """It should return True for a string with special characters."""
+        self.assertTrue(is_unique("!@#$%^&*()"))
+
+    # Defensive tests
+    def test_not_string_input(self):
```
Comment (Urz1) [inline solutions/tests/test_is_unique.py:54]: While you are raising AssertionError for invalid inputs, you can improve the tests by asserting that the error message matches the expected one.
```suggestion
    ```
def test_not_string_input(self):
    """It should raise AssertionError for non-string input with the correct error message."""
       with self.assertRaises(AssertionError) as context:
        is_unique(123)
    self.assertEqual(str(context.exception), "Input must be a string")
```

```

**Comment by Urz1** [comment]
Great job on writing clear and comprehensive solution with it's tests! I have checked it all it works well. I have added some potential improvement areas check it out. Kudos to you

**Review by Urz1 — CHANGES_REQUESTED** [review]
Great Job!

**Review by Urz1 — APPROVED** [review]

File: solutions/tests/test_is_unique.py (hunk 1 above)
Comment (ciiyaa) [reply solutions/tests/test_is_unique.py:46]: Thanks for this suggestion!

File: solutions/tests/test_is_unique.py (hunk 2 above)
Comment (ciiyaa) [reply solutions/tests/test_is_unique.py:54]: I'm not sure if I understand well, could you elaborate more.

Comment (Urz1) [reply solutions/tests/test_is_unique.py:54]: It's just a suggestion to improve on the messaging. you can leave it it's normal this way as well.

**Review by Urz1 — APPROVED** [review]
Great work!
The code aligns well with the defined criteria, including proper documentation, defensive assertions, and comprehensive test coverage for standard, edge, and boundary cases. The addition of helpful docstrings and clean formatting enhances readability and maintainability.
Tests run successfully, and all logic is simple and efficient. Approved for merge! 🎉

Screenshot from 2025-01-04 19-35-19
